contrib/IECoreAlembic/src/IECoreAlembic/GeomBaseAlgo.cpp

```diff
@@ -0,0 +1,238 @@
+
+#include "IECore/MessageHandler.h"
+
+#include "IECoreAlembic/GeomBaseAlgo.h"
+#include "IECoreAlembic/IGeomParamTraits.h"
+
+using namespace Alembic::Abc;
+using namespace Alembic::AbcGeom;
+using namespace IECore;
+using namespace IECoreAlembic;
+
+//////////////////////////////////////////////////////////////////////////
+// Internal utilities
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+IECore::PrimitiveVariable::Interpolation interpolationFromScope( Alembic::AbcGeom::GeometryScope scope )
+{
+	switch( scope )
+	{
+		case kConstantScope :
+			return PrimitiveVariable::Constant;
+		case kUniformScope :
+			return PrimitiveVariable::Uniform;
+		case kVaryingScope :
+			return PrimitiveVariable::Varying;
+		case kVertexScope :
+			return PrimitiveVariable::Vertex;
+		case kFacevaryingScope :
+			return PrimitiveVariable::FaceVarying;
+		default :
+			return PrimitiveVariable::Invalid;
+	}
+}
+
+
+template<typename DataType, typename GeomParam>
+struct ApplyGeometricInterpretation
+{
+
+	static void apply( DataType *data )
+	{
+	};
+
+};
+
+template<typename T, typename GeomParam>
+struct ApplyGeometricInterpretation<GeometricTypedData<T>, GeomParam>
+{
+
+	static void apply( GeometricTypedData<T> *data )
+	{
+		data->setInterpretation( IGeomParamTraits<GeomParam>::geometricInterpretation() );
+	};
+
+};
+
+} // namespace
+
+//////////////////////////////////////////////////////////////////////////
+// GeomBaseAlgo implementation
+//////////////////////////////////////////////////////////////////////////
+
+namespace IECoreAlembic
+{
+
+namespace GeomBaseAlgo
+{
+
+void convertUVs( const Alembic::AbcGeom::IV2fGeomParam &uvs, const Alembic::Abc::ISampleSelector &sampleSelector, IECore::Primitive *primitive )
+{
+	if( !uvs.valid() )
+	{
+		return;
+	}
+
+	/// \todo It'd be nice if we stored uvs as a single primitive variable instead of having to split them in two.
+	/// It'd also be nice if we supported indexed data directly.
+	typedef IV2fArrayProperty::sample_ptr_type SamplePtr;
+	SamplePtr sample = uvs.getExpandedValue( sampleSelector ).getVals();
+	size_t size = sample->size();
+
+	FloatVectorDataPtr sData = new FloatVectorData;
+	FloatVectorDataPtr tData = new FloatVectorData;
+	std::vector<float> &s = sData->writable();
+	std::vector<float> &t = tData->writable();
+	s.resize( size );
+	t.resize( size );
+	for( size_t i=0; i<size; ++i )
+	{
+		s[i] = (*sample)[i][0];
+		t[i] = (*sample)[i][1];
+	}
+
+	PrimitiveVariable::Interpolation interpolation = interpolationFromScope( uvs.getScope() );
+	primitive->variables["s"] = PrimitiveVariable( interpolation, sData );
+	primitive->variables["t"] = PrimitiveVariable( interpolation, tData );
+}
+
+void convertArbGeomParams( const Alembic::Abc::ICompoundProperty &params, const Alembic::Abc::ISampleSelector &sampleSelector, IECore::Primitive *primitive )
+{
+	if( !params.valid() )
```

Comment:
I'm no expert here, but I don't _think_ so. I think `!valid()` is just saying that no arbGeomParams have been written, not that there is actually an error. Alembic's own Arnold procedural certainly seems to [treat it that way](https://github.com/alembic/alembic/blob/master/arnold/Procedural/ArbGeomParams.cpp#L331). If I do raise here instead then the tests fail, because the test files do include objects without arbGeomParams.

In practice, I think what we have here already is probably what we want - we want a function that applies any arbGeomParams if they exist, and otherwise doesn't cause trouble.

Actually, it seems like it might be nice to take an `const IGeomBase &` argument instead of the `ICompoundProperty &`, so we can enforce that the property we work on is actually expected to hold arbgeomparams. But as far as I can tell things like `ICurves` and `ISubD` don't actually derive from `IGeomBase`, so this doesn't work. Or I guess we could template it, but I'm not sure the benefits really warrant the extra fuss.